tasmota/xsns_92_hdc1080.ino

```diff
@@ -0,0 +1,288 @@
+ *
+ * Source: Luis Teixeira
+ *
+ * I2C Address: 0x40
+\*********************************************************************************************/
+
+#define XSNS_92             92
+#define XI2C_92             92      // See I2CDEVICES.md
+
+#define HDC1080_ADDR        0x40
+
+// Registers:
+
+#define HDC_REG_TEMP        0x00    // Temperature register
+#define HDC_REG_RH          0x01    // Humidity register
+#define HDC_REG_CONFIG      0x02    // Configuration register
+#define HDC_REG_SERIAL1     0xFB    // First 2 bytes of the serial ID
+#define HDC_REG_SERIAL2     0xFC    // Mid 2 bytes of the serial ID
+#define HDC_REG_SERIAL3     0xFD    // Last bits of the serial ID
+#define HDC_REG_MAN_ID      0xFE    // Manufacturer ID
+#define HDC_REG_DEV_ID      0xFF    // Device ID
+
+// Expected constant values of some of the registers:
+
+#define HDC1080_MAN_ID      0x5449  // Manufacturer ID (Texas Instruments)
+#define HDC1080_DEV_ID      0x1050  // Device ID (valid for the HDC1080)
+
+// Possible values for the configuration register fields:
+
+#define HDC1080_RST_ON      0x8000
+#define HDC1080_HEAT_ON     0x2000
+#define HDC1080_MODE_ON     0x1000  // acquision mode (temperature + humidity)
+#define HDC1080_TRES_11     0x400
+#define HDC1080_HRES_11     0x100
+#define HDC1080_HRES_8      0x80
+
+// Constants:
+
+#define HDC1080_CONV_TIME   50      // Assume 6.50 + 6.35 ms + x of conversion delay for this device
+#define HDC1080_TEMP_MULT   0.0025177
+#define HDC1080_RH_MULT     0.0025177
+#define HDC1080_TEMP_OFFSET 40.0
+
+
+const char* hdc_type_name = "HDC1080";
+uint16_t hdc_manufacturer_id = 0;
+uint16_t hdc_device_id = 0;
+
+float hdc_temperature = 0.0;
+float hdc_humidity = 0.0;
+
+uint8_t hdc_valid = 0;
+
+/**
+ * Reads the device ID register.
+ * 
+ */
+uint16_t HdcReadDeviceId(void) {
+  return I2cRead16(HDC1080_ADDR, HDC_REG_DEV_ID);
+}
+
+/**
+ * Reads the manufacturer ID register.
+ * 
+ */
+uint16_t HdcReadManufacturerId(void) {
+  return I2cRead16(HDC1080_ADDR, HDC_REG_MAN_ID);
+}
+
+/**
+ * Overwrites the configuration register with the provided config
+ */
+void HdcConfig(uint16_t config) {
+  I2cWrite16(HDC1080_ADDR, HDC_REG_CONFIG, config);
+}
+
+/**
+ * Performs a soft reset on the device.
+ * 
+ * RST = 1 -> software reset
+ * 
+ */
+void HdcReset(void) {
+  uint16_t current = I2cRead16(HDC1080_ADDR, HDC_REG_CONFIG);
+  
+  // bit 15 of the configuration register contains the RST flag
+  // so we set it to 1:
+
+  current |= 0x8000;
+
+  I2cWrite16(HDC1080_ADDR, HDC_REG_CONFIG, current);
+
+  delay(HDC1080_CONV_TIME);  // Not sure how long it takes to reset. Assuming this value.
+}
+
+/**
+ * Performs the single transaction read of the HDC1080, providing the 
+ * adequate delay for the acquisition.
+ * 
+ */
+int8_t HdcReadBuffer(uint8_t addr, uint8_t reg, uint8_t *reg_data, uint16_t len) {
+  Wire.beginTransmission((uint8_t)addr);
+  Wire.write((uint8_t)reg);
+  Wire.endTransmission();
+
+  delay(HDC1080_CONV_TIME);
```

Comment:
Thanks for the PR. Most of the time `delay()` is to be avoided since you're stopping the whole Tasmota for a full 50ms cycle. Could you just let it run and read the value at next 50ms tick, if possible?

Reply:
Hi @s-hadinger, understood. I'll take a look at that and test. Also, 50 ms is a conservative time which I kept while I was troubleshooting the communication. Theoretically 15 ms should be sufficient, according to the datasheet anyway.

Reply:
@s-hadinger do you have a good example where we are doing this?  I am pondering on using FUNC_EVERY_SECOND as long as the sensor doesn't care that we read the values that long after sending the command. Apart from that it should not be a problem as our sample rate in Tasmota doesn't go below that anyway.

Reply:
The way you could is simply send the command to the sensor, use a boolean to know that you expect a reading at next tick, and return.
Then use `FUNC_EVERY_50_MSECOND`. At the next FUNC_EVERY_50_MSECOND call, just read back the value and reset the boolean. You could argue that there might be less than 50ms between the initial call and the read request. If you want to protect, just set a `uint32_t timer = millis() + 30;`. At each `FUNC_EVERY_50_MSECOND`, test using `TimeReached(timer)`. If true, do the reading and reset the timer, if false return and wait for next tick.